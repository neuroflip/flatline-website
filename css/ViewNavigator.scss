.viewNavigatorContainer {
    user-select: none;
    -webkit-user-drag: none;
    position: fixed;
    top: 72px;
    left: 0px;
    width: 100%;
    height: 100%;
    padding-left: 15px;
    //max-width: 500px;
}