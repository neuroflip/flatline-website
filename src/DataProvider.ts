import Column1Data from './data/Crawlear.html'
import Column2Data from './data/ObsoleteEngine.html'
import Column3Data from './data/Tunnel.html'
import Column4Data from './data/Pulgon.html'
import Column5Data from './data/SpaceTrash.html'
import Column6Data from './data/Tetris.html'
import Column7Data from './data/AMP.html'

const mod = (x: number, y: number): number => {
  return ((x % y) + y) % y
}

class DataProvider {
  data: any[]

  constructor () {
    this.data = [Column1Data,
      Column2Data,
      Column3Data,
      Column4Data,
      Column5Data,
      Column6Data,
      Column7Data]
  }

  getCurrent (index: number): TrustedHTML {
    return this.data[mod(index, this.data.length)]
  }

  getNext (index: number): TrustedHTML {
    const current = mod((index + 1), this.data.length)

    return this.data[current]
  }

  getPrevious (index: number): TrustedHTML {
    const current = index - 1

    return this.getCurrent(current)
  }
}

export default DataProvider
