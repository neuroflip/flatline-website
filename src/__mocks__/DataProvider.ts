const DataProvider = jest.fn().mockImplementation(() => {
  return {
    getCurrent: jest.fn().mockImplementation((index: number): TrustedHTML => {
      if (index === 0) {
        return `<div>column1 ${index}</div>`
      } else {
        return `<div>column2 ${index}</div>`
      }
    }),
    getNext: jest.fn().mockImplementation((index: number): TrustedHTML => {
      if (index === 0) {
        return `<div>column2 ${index}</div>`
      } else {
        return `<div>column3 ${index}</div>`
      }
    }),
    getPrevious: jest.fn().mockImplementation((index: number): TrustedHTML => {
      if (index === 0) {
        return `<div>column5 ${index}</div>`
      } else {
        return `<div>column1 ${index}</div>`
      }
    })
  }
})

export default DataProvider
