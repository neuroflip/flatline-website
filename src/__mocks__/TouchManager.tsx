import * as React from 'react'
import type { TouchManagerProps } from '../types/TouchManager'

const TouchManager = ({ onStartTouch, onMoveTouch, onPageChange, onResetColumns, children }: TouchManagerProps): React.ReactElement => {
  React.useEffect(() => {
    addEvents()
  }, [])

  function addEvents (): void {
    const root = document.getElementById('root') ?? window.document.body

    root.addEventListener('mousedown', () => {
      onStartTouch()
    })
    root.addEventListener('mousemove', () => {
      onMoveTouch(-100)
    })
    root.addEventListener('mouseup', () => {
      onPageChange(-400)
    })
  }

  return children
}
export default TouchManager
