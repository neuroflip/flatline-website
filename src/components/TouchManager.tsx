import * as React from 'react'
import Utils from '../Utils'
import type { TouchManagerProps } from '../types/TouchManager'

const body = window.document.body
const MIN_SENSI = 100

function TouchManager ({
  onPageChange,
  onStartTouch,
  onMoveTouch,
  onResetColumns,
  children
}: TouchManagerProps): React.JSX.Element {
  const stateRef = React.useRef({ startX: 0, startY: 0, isRealEvent: false })

  React.useEffect(() => {
    addEvents()

    return () => {
      removeEvents()
    }
  }, [])

  function addEvents (): void {
    body.addEventListener(Utils.isMobile() ? 'touchstart' : 'mousedown', onTouchStart)
    body.addEventListener(Utils.isMobile() ? 'touchmove' : 'mousemove', onTouchMove)
    body.addEventListener(Utils.isMobile() ? 'touchend' : 'mouseup', onTouchEnd)
    window.addEventListener('resize', onResetColumns())
  }

  function removeEvents (): void {
    body.removeEventListener(Utils.isMobile() ? 'touchstart' : 'mousedown', onTouchStart)
    body.removeEventListener(Utils.isMobile() ? 'touchmove' : 'mousemove', onTouchMove)
    body.removeEventListener(Utils.isMobile() ? 'touchend' : 'mouseup', onTouchEnd)

    window.removeEventListener('resize', onResetColumns())
  }

  function onTouchStart (nativeEvent: TouchEvent | MouseEvent): void {
    if (nativeEvent instanceof MouseEvent) {
      stateRef.current.startX = nativeEvent.clientX
      stateRef.current.startY = nativeEvent.clientY
    } else if (nativeEvent instanceof TouchEvent) {
      const touches = nativeEvent.touches

      if (Boolean(touches) && touches.length === 1) {
        stateRef.current.startX = touches[0].clientX
        stateRef.current.startY = touches[0].clientY
      }
    }
    stateRef.current.isRealEvent = true
    onStartTouch()
  }

  function onTouchMove (nativeEvent: TouchEvent | MouseEvent): void {
    let progressX = 0
    let progressY = 0

    if (stateRef.current.isRealEvent) {
      if (nativeEvent instanceof MouseEvent) {
        progressX = nativeEvent.clientX
        progressY = nativeEvent.clientY
      } else if (nativeEvent instanceof TouchEvent) {
        progressX = nativeEvent.touches[0].clientX
        progressY = nativeEvent.touches[0].clientY
      }
      progressX = Math.round(progressX) - stateRef.current.startX
      progressY = Math.round(progressY) - stateRef.current.startY

      if (Math.abs(progressX) > Math.abs(progressY)) {
        onMoveTouch(progressX)
      }
    }
  }

  function onTouchEnd (nativeEvent: TouchEvent | MouseEvent): void {
    let finalProgressX = 0
    let finalProgressY = 0

    if (nativeEvent instanceof MouseEvent) {
      finalProgressX = nativeEvent.clientX
      finalProgressY = nativeEvent.clientY
    } else if (nativeEvent instanceof TouchEvent) {
      finalProgressX = nativeEvent.changedTouches[0].clientX
      finalProgressY = nativeEvent.changedTouches[0].clientY
    }

    finalProgressX = (finalProgressX) - stateRef.current.startX
    finalProgressY = (finalProgressY) - stateRef.current.startY
    stateRef.current.isRealEvent = false

    if (Math.abs(finalProgressX) > MIN_SENSI && Math.abs(finalProgressX) > Math.abs(finalProgressY)) {
      onPageChange(finalProgressX)
    } else if (finalProgressX !== 0) {
      onResetColumns()
    }
  }

  return children
}

export default TouchManager
