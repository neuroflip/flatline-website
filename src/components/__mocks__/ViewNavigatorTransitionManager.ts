import type { ViewNavigatorState } from '../../types/ViewNavigator'

const ViewNavigatorTransitionManager = jest.fn().mockImplementation(() => {
  return {
    constructor: jest.fn().mockImplementation(()=>{
      console.log('init! constructor en mock de transitioner')
    }),
    setInitialColumnWidth: jest.fn().mockImplementation((state: ViewNavigatorState, width: number): ViewNavigatorState => {
      const newState: ViewNavigatorState = {
        index: 0,
        lastMove: 0,
        col1Classes: '',
        col2Classes: 'current',
        col3Classes: '',
        columns: [],
        col1Style: { transform: 'translate(0px, 0px)' },
        col2Style: { transform: 'translate(0px, 0px)' },
        col3Style: { transform: 'translate(0px, 0px)' }
      }

      return newState
    }),
    removeAnimation: jest.fn().mockImplementation((state: ViewNavigatorState): ViewNavigatorState => {
      const newState: ViewNavigatorState = {
        index: 0,
        lastMove: 0,
        col1Classes: '',
        col2Classes: 'current',
        col3Classes: '',
        columns: [],
        col1Style: { transform: 'translate(0px, 0px)' },
        col2Style: { transform: 'translate(0px, 0px)' },
        col3Style: { transform: 'translate(0px, 0px)' }
      }

      return newState
    }),
    onMovePage: jest.fn().mockImplementation((state: ViewNavigatorState, progressX: number, columnWidth: number): ViewNavigatorState => {
      const newState: ViewNavigatorState = {
        index: 0,
        lastMove: 0,
        col1Classes: '',
        col2Classes: '',
        col3Classes: 'current',
        columns: [],
        col1Style: { transform: 'translate(-300px, 0px)' },
        col2Style: { transform: 'translate(0px, 0px)' },
        col3Style: { transform: 'translate(300px, 0px)' }
      }

      return newState
    }),
    onResetColumns: jest.fn().mockImplementation((state: ViewNavigatorState, columnWidth: number): ViewNavigatorState => {
      const newState: ViewNavigatorState = {
        index: 0,
        lastMove: 0,
        col1Classes: '',
        col2Classes: 'current',
        col3Classes: '',
        columns: [],
        col1Style: { transform: 'translate(0px, 0px)' },
        col2Style: { transform: 'translate(0px, 0px)' },
        col3Style: { transform: 'translate(0px, 0px)' }
      }

      return newState
    }),
    onPageChange: jest.fn().mockImplementation((state: ViewNavigatorState, finalProgressX: number, columnWidth: number): ViewNavigatorState => {
      const newState: ViewNavigatorState = {
        index: 1,
        lastMove: 0,
        col1Classes: '',
        col2Classes: '',
        col3Classes: 'current',
        columns: [],
        col1Style: { transform: 'translate(300px, 0px)' },
        col2Style: { transform: 'translate(-300px, 0px)' },
        col3Style: { transform: 'translate(0px, 0px)' }
      }

      return newState
    })

  }
})

export default ViewNavigatorTransitionManager
