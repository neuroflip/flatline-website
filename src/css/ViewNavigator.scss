.viewNavigatorContainer {
    user-select: none;
    -webkit-user-select: none;
    -webkit-user-drag: none;
    width: 93%;
    height: 100%;
    margin: 0 auto;
}

@media only screen and (min-width : 400px) {
    .viewNavigatorContainer {
        width: 70%;
    }
}
