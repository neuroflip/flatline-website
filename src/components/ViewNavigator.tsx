import * as React from 'react'
import InputManager from './InputManager'
import Column from './Column'
import ArrowHandler from './ArrowHandler'
import useViewNavigator from '../hooks/useViewNavigator'

import '../css/ViewNavigator.scss'
import { ViewNavigatorProps } from '../types/ViewNavigator'

function ViewNavigator ({ onClose }: ViewNavigatorProps): React.JSX.Element {
  const [state, col1, col2, col3, onPageChange, onMovePage, onResetColumns, onArrowLeftClick, onArrowRightClick] = useViewNavigator()

  return <InputManager onPageChange={onPageChange}
    onStartTouch={() => {}}
    onMovePage={onMovePage}
    onResetColumns={onResetColumns}>
      <>
        <div className='closeButton' onClick={onClose}>x</div>
        <div className='viewNavigatorContainer'>
            <Column ref={state.columns[0]} id='col1' className={`col1 ${state.col1Classes}`} style={state.col1Style}>
              <div className="columnContainer" dangerouslySetInnerHTML={{ __html: col1 }} />
            </Column>
            <Column ref={state.columns[1]} id='col2' className={`col2 ${state.col2Classes}`} style={state.col2Style}>
              <div className="columnContainer" dangerouslySetInnerHTML={{ __html: col2 }} />
            </Column>
            <Column ref={state.columns[2]} id='col3' className={`col3 ${state.col3Classes}`} style={state.col3Style}>
              <div className="columnContainer" dangerouslySetInnerHTML={{ __html: col3 }} />
            </Column>
        </div>
        <ArrowHandler onRightClick={onArrowRightClick} onLeftClick={onArrowLeftClick}></ArrowHandler>
      </>
  </InputManager>
}

export default ViewNavigator
