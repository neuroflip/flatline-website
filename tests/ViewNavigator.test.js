/**
 * @jest-environment jsdom
 */

import * as React from 'react'
import { render, screen, fireEvent } from '@testing-library/react'
import renderer from 'react-test-renderer'
import ViewNavigator from '../src/components/ViewNavigator'
import DataProvider from '../src/DataProvider'
import TouchManager from '../src/TouchManager'
import ViewNavigatorTransitionManager from '../src/components/ViewNavigatorTransitionManager'
import '@testing-library/jest-dom'

let container = null

jest.mock('../src/TouchManager')
jest.mock('../src/DataProvider')
jest.mock('../src/components/ViewNavigatorTransitionManager')

beforeEach(() => {
  DataProvider.mockClear()

  container = window.document.createElement('div')
  container.id = 'root'
  window.document.body.appendChild(container)
})

afterEach(() => {
  window.document.body.removeChild(container)
})

it('ViewNavigator loads the first content in columns', () => {
  render(<ViewNavigator></ViewNavigator>, container)

  expect(screen.queryByText('column1 0')).toBeInTheDocument()
  expect(screen.queryByText('column2 0')).toBeInTheDocument()
  expect(screen.queryByText('column5 0')).toBeInTheDocument()
  expect(screen.queryByText('column3 0')).not.toBeInTheDocument()
  expect(screen.queryByText('column4 0')).not.toBeInTheDocument()
})

it('ViewNavigator loads the next content in columns', () => {
  render(<ViewNavigator></ViewNavigator>, container)

  fireEvent.mouseDown(container, {
    target: container,
    clientX: 10,
    clientY: 100
  })

  fireEvent.mouseMove(container, {
    target: container,
    clientX: -300,
    clientY: 100
  })

  fireEvent.mouseUp(container, {
    target: container,
    clientX: -300,
    clientY: 100
  })

  expect(screen.queryByText('column1 1')).toBeInTheDocument()
  expect(screen.queryByText('column2 1')).toBeInTheDocument()
  expect(screen.queryByText('column3 1')).toBeInTheDocument()
})

it('snapshot test', () => {
  const component = renderer.create(<ViewNavigator></ViewNavigator>)
  const tree = component.toJSON()

  expect(tree).toMatchSnapshot()
})

