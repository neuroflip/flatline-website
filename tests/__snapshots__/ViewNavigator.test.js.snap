// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test 1`] = `
[
  <div
    className="viewNavigatorContainer"
  >
    <div
      className="col1 "
      id="col1"
      style={
        {
          "transform": "translate(-0px,0) perspective(400px) rotateY(-45deg) scale(0.5)",
        }
      }
    >
      <div
        className="columnContainer"
        dangerouslySetInnerHTML={
          {
            "__html": "<div>column5 0</div>",
          }
        }
      />
    </div>
    <div
      className="col2 current"
      id="col2"
      style={
        {
          "transform": "translate(0px, 0px)",
        }
      }
    >
      <div
        className="columnContainer"
        dangerouslySetInnerHTML={
          {
            "__html": "<div>column1 0</div>",
          }
        }
      />
    </div>
    <div
      className="col3 "
      id="col3"
      style={
        {
          "transform": "translate(0px,0) perspective(400px) rotateY(-45deg) scale(0.5)",
        }
      }
    >
      <div
        className="columnContainer"
        dangerouslySetInnerHTML={
          {
            "__html": "<div>column2 0</div>",
          }
        }
      />
    </div>
  </div>,
  <div
    className="left arrowHandler"
    id="leftArrowHandler"
    onClick={[Function]}
  >
    &lt;
  </div>,
  <div
    className="right arrowHandler"
    id="rigthArrowHandler"
    onClick={[Function]}
  >
    &gt;
  </div>,
]
`;
